Sources/WhisperKit/Core/TextDecoder.swift

```diff
@@ -308,14 +316,100 @@ public class TextDecoder: TextDecoding, WhisperMLModel {
+            throw WhisperError.modelsUnavailable("Failed to read logits size from model")
+        }
+
+        // Logits filters
+        var logitsFilters: [any LogitsFiltering] = []
+        let allLanguageTokens:[Int] = tokenizer.languages.compactMap {tokenizer.convertTokenToId("<|\($0)|>")}
+        
+        // language filter
+        logitsFilters.append(
+            LanguageLogitsFilter(
+                allLanguageTokens: allLanguageTokens,
+                logitsDim: logitsSize,
+                sampleBegin: prefilledIndex)
+        )
+        Logging.debug("prefilled Index is \(prefilledIndex)")
+
+        let tokenIndex = 0
+        let prefillToken = currentTokens[tokenIndex]
+        nextToken = prefillToken
+        Logging.debug("Forcing token \(nextToken) at index \(tokenIndex) from initial prompt")
+        
+        
+        // Set the current token as model input
+        decoderInputs.inputIds[0] = NSNumber(value: nextToken)
+        decoderInputs.cacheLength[0] = NSNumber(value: tokenIndex)
+        
+        // MARK: Decoding Inference
+        
+        // Predict next token
+        let inferenceTime = Date()
+        
+        let predictedLogits = try await self.predictLogits(
+            inputIds: decoderInputs.inputIds,
+            cacheLength: decoderInputs.cacheLength,
+            keyCache: decoderInputs.keyCache,
+            valueCache: decoderInputs.valueCache,
+            kvCacheUpdateMask: decoderInputs.kvCacheUpdateMask,
+            encoderOutputEmbeds: encoderOutput,
+            decoderKeyPaddingMask: decoderInputs.decoderKeyPaddingMask
+        )
+        Logging.debug("Predicted logits")
+        guard let decoderOutput = predictedLogits else {
+            Logging.error("Unable to decode logits")
+            throw WhisperError.decodingLogitsFailed()
+        }
+        
+        let decodingInferenceTime = Date().timeIntervalSince(inferenceTime)
+        timings.decodingPredictions += decodingInferenceTime
+        
+        // MARK: Non-inference
+        
+        // Update predicted token as current
+        var logits = decoderOutput.logits!
+        for filter in logitsFilters {
+            logits = filter.filterLogits(logits, withTokens: currentTokens)
+        }
+        
+        // MARK: Sampling
+        
+        let samplingStartTime = Date()
+        
+        let sampleResult = tokenSampler.update(tokens: currentTokens, logits: logits, logProbs: logProbs)
+        
+        nextToken = sampleResult.tokens.last!
+        logProbs = sampleResult.logProbs
+        
+        let samplingTime = Date().timeIntervalSince(samplingStartTime)
+        timings.decodingSampling += samplingTime
+        
+        var detectedLanguage = tokenizer.decode(tokens: [nextToken]).dropFirst(2).dropLast(2)
+        var decodingResult = DecodingResult.emptyResults
+        decodingResult.timings = timings
+        decodingResult.language = String(detectedLanguage)
+        return [decodingResult]
```

Comment:
There's quite a bit of repeated code here that may get out of date with the main decoder loop, if you're up for it would be great to see an attempt at unifying them, but won't block merge.

Reply:
Let me do it in a seperate PR ?

Reply:
Sure 👍